src/box/box.cc

```diff
@@ -2246,8 +2246,38 @@ box_set_bootstrap_leader_record(void)
 }
 
 int
-box_make_bootstrap_leader(void)
+box_make_bootstrap_leader(bool graceful)
 {
+	/* Bootstrap strategy is read by the time instance uuid is known. */
+	if (!tt_uuid_is_nil(&INSTANCE_UUID) &&
+	    bootstrap_strategy != BOOTSTRAP_STRATEGY_SUPERVISED) {
+		assert(bootstrap_strategy != BOOTSTRAP_STRATEGY_INVALID);
+		diag_set(ClientError, ER_UNSUPPORTED,
+			 tt_sprintf("bootstrap_strategy = '%s'",
+				    cfg_gets("bootstrap_strategy")),
+			 "promoting the bootstrap leader via "
+			 "box.ctl.make_bootstrap_leader()");
+		return -1;
+	}
+
+	if (graceful && is_box_configured) {
+		is_graceful_supervised_bootstrap_requested = false;
+		say_info("graceful bootstrap request is discarded: the "
+			 "instance is already bootstrapped");
+		return 0;
+	} else if (graceful) {
+		is_graceful_supervised_bootstrap_requested = true;
+		say_info("graceful bootstrap is requested: the instance "
+			 "is going to check whether a bootstrap leader "
+			 "already exists and if the check fails the instance "
+			 "takes this role");
+		if (!tt_uuid_is_nil(&INSTANCE_UUID))
+			box_broadcast_ballot();
```

Comment:
Why do you need a `box_broadcast_ballot()` here?

Reply:
Added a short remark:

```c
/* Wake up replicaset_connect() if the flag is turn on. */
if (!tt_uuid_is_nil(&INSTANCE_UUID))
        box_broadcast_ballot();
```